【Hackathon 5th No.26】为 Paddle 新增 diagonal_scatter API -part

DanGuge opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
Others
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
Others
### Description
<!-- Describe what you’ve done -->
为 Paddle 新增 diagonal_scatter API
RFC：[PaddlePaddle/community#669](https://github.com/PaddlePaddle/community/pull/669)
#57262

DanGuge committed:
add diagonal_scatter

DanGuge committed:
add diagonal_scatter_test

DanGuge committed:
add annotations

DanGuge committed:
modify int32/int64 test range

DanGuge committed:
update diagonal_scatter doc

DanGuge commented:
@zxcd 辛苦review一下，diagonal_scatter的实现

zxcd commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5575,140 @@ def index_fill_(x, index, axis, value, name=None):
+                'float16',
+                'float32',
+                'float64',
+                'bfloat16',
```
bfloat16目前在这里应该是使用uint16?

zxcd commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5575,140 @@ def index_fill_(x, index, axis, value, name=None):
+    if in_dynamic_mode():
+        return _C_ops.fill_diagonal_tensor(x, y, offset, axis1, axis2)
+    else:
+        helper = LayerHelper('diagonal_scatter', **locals())
```
建议放在check_variable_and_dtype后面

zxcd commented on test/legacy_test/test_diagonal_scatter.py:
```diff
@@ -0,0 +1,353 @@
+
+# check error
+class TestDiagonalScatterError(TestDiagonalScatterAPI):
+    def test_error_1(self):
```
改成有意义的命名

DanGuge committed:
fix: bfloat16 type & test name & LayerHelper position

DanGuge commented:
@zxcd 修改完了，辛苦再review一下

zxcd reviewed (approved):
LGTM

jeff41404 commented on python/paddle/__init__.py:
```diff
@@ -256,6 +256,7 @@
     masked_fill_,
     index_fill,
     index_fill_,
+    diagonal_scatter,
```
need to be added in list of \_\_all\_\_ below

jeff41404 commented:
The logic of the `diagonal_scatter` API and the existing function `fill_diagonal_tensor` in `python/paddle/tensor/manipulation.py` are most of the same, whether it can be directly exposed as an alias of `fill_diagonal_tensor` function, so no need to add additional Python logic? Meanwhile, its underlying operator `fill_ Diagonal_ tensor` needs to add support for `int16` data types

DanGuge committed:
add diagonal scatter in __all__ list

DanGuge commented:
> The logic of the `diagonal_scatter` API and the existing function `fill_diagonal_tensor` in `python/paddle/tensor/manipulation.py` are most of the same, whether it can be directly exposed as an alias of `fill_diagonal_tensor` function, so no need to add additional Python logic? Meanwhile, its underlying operator `fill_ Diagonal_ tensor` needs to add support for `int16` data types

@jeff41404 
1. There is still difference between `fill_diagonal_tensor` and `diagonal_scatter`, the former one is only available in **dynamic mode** while the latter one supports both **dynamic & static mode**. And I think maybe we should not change the origin usage of `fill_diagonal_tensor`, so it may be better to add additional API for `diagonal_scatter`.
2. I can submit another PR for `int16` data types of `fill_diagonal_tensor` & add relative changes, e.g. tests and so on.

jeff41404 commented:
> 1. There is still difference between `fill_diagonal_tensor` and `diagonal_scatter`, the former one is only available in **dynamic mode** while the latter one supports both **dynamic & static mode**. And I think maybe we should not change the origin usage of `fill_diagonal_tensor`, so it may be better to add additional API for `diagonal_scatter`.

Adding branch of static mode to the current API `fill_diagonal_tensor` is the compatibility upgrade and correct approach, which can reduce future maintenance costs, as long as it is clearly described in the rfc document and pass existing test case.

jeff41404 commented:
> 2\. I can submit another PR for `int16` data types of `fill_diagonal_tensor` & add relative changes, e.g. tests and so on.

OK, you can add link of PR to the description above and add these tasks to rfc

DanGuge committed:
add int16 in fill_diagonal_tensor & reuse the logic of fill_diagonal_tensor for diagonal_scatter

DanGuge committed:
remove xpu settings

DanGuge committed:
op test get_numeric_gradient does not support int16

DanGuge committed:
fix

DanGuge commented:
@jeff41404 cc @zxcd 
I added the logic of static mode to `fill_diagonal_tensor` and simplified `diagonal_scatter`.
And I added `int16` to `fill_diagonal_tensor` in this PR, please review it.

jeff41404 reviewed (dismissed):
LGTM

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
+    This function returns a tensor with fresh storage.
+
+    The argument ``offset`` controls which diagonal to consider:
+    If ``offset`` = 0, it is the main diagonal.
```
加上 - 更直观，下同

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
+        ``y`` should have the same shape as paddle.diagonal(x, offset, axis1, axis2).
+
+    Args:
+        x (Tensor): `x`` is the original Tensor. Must be at least 2-dimensional.
```
```suggestion
        x (Tensor): ``x`` is the original Tensor. Must be at least 2-dimensional.
```

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
+        name (str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor, Tensor with diagonal embedeed with y.
```
```suggestion
        Tensor, Tensor with diagonal embedeed with ``y``.
```

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
+    Examples:
+        .. code-block:: python
+
+            >>> # doctest: +REQUIRES(env:GPU)
```
这个是一定需要 GPU吗，如果不是，改成使用CPU的示例代码比较好~

DanGuge committed:
Update python/paddle/tensor/manipulation.py

Co-authored-by: ooo oo <106524776+ooooo-create@users.noreply.github.com>

DanGuge committed:
Update python/paddle/tensor/manipulation.py

Co-authored-by: ooo oo <106524776+ooooo-create@users.noreply.github.com>

DanGuge committed:
fix review suggestions

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5770,3 +5827,45 @@ def index_fill_(x, index, axis, value, name=None):
+
+    This function returns a tensor with fresh storage.
+
+    The argument ``offset`` controls which diagonal to consider:
```
下面加个空行~

ooooo-create commented on python/paddle/tensor/manipulation.py:
```diff
@@ -5770,3 +5827,45 @@ def index_fill_(x, index, axis, value, name=None):
+    - If ``offset`` < 0, it is below the main diagonal.
+
+    Note:
+        ``y`` should have the same shape as paddle.diagonal(x, offset, axis1, axis2).
```
写成这样会不会更好~
```suggestion
        ``y`` should have the same shape as :ref:`paddle.diagonal <api_paddle_diagonal>`.
```

ooooo-create reviewed (commented):
其余LGTM @sunzhongkai588

DanGuge committed:
Update python/paddle/tensor/manipulation.py

Co-authored-by: ooo oo <106524776+ooooo-create@users.noreply.github.com>

DanGuge committed:
fix

ooooo-create reviewed (approved):
LGTM @sunzhongkai588

sunzhongkai588 reviewed (approved):
LGTM for docs

luotao1 commented:
hi, @DanGuge 
- 非常感谢你对飞桨框架的贡献，我们正在运营一个PFCC组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨框架做贡献，详情可见 https://github.com/luotao1 主页说明。
- 如果你对PFCC有兴趣，请发送邮件至 [ext_paddle_oss@baidu.com](mailto:ext_paddle_oss@baidu.com)，我们会邀请你加入~
